File: src/main/java/com/skysoft/mixin/MinecraftSlotLockMixin.java
package com.skysoft.mixin;

import com.llamalad7.mixinextras.injector.wrapoperation.Operation;
import com.llamalad7.mixinextras.injector.wrapoperation.WrapOperation;
import com.skysoft.features.inventory.SlotLockManager;
import com.skysoft.utils.input.InputHandlingResult;
import net.minecraft.client.Minecraft;
import net.minecraft.client.multiplayer.ClientPacketListener;
import net.minecraft.network.protocol.Packet;
import net.minecraft.network.protocol.game.ServerboundPlayerActionPacket;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;

@Mixin(Minecraft.class)
public class MinecraftSlotLockMixin {
    @WrapOperation(
        method = "handleKeybinds",
        at = @At(
            value = "INVOKE",
            target = "Lnet/minecraft/client/multiplayer/ClientPacketListener;send(Lnet/minecraft/network/protocol/Packet;)V"
        )
    )
    private void skysoft$protectLockedSlotsFromOffhandSwap(
        ClientPacketListener connection,
        Packet<?> packet,
        Operation<Void> original
    ) {
        Minecraft minecraft = (Minecraft) (Object) this;
        if (
            packet instanceof ServerboundPlayerActionPacket action
                && action.getAction() == ServerboundPlayerActionPacket.Action.SWAP_ITEM_WITH_OFFHAND
                && SlotLockManager.handleOffhandSwap(minecraft.player) == InputHandlingResult.CONSUMED
        ) {
            return;
        }
        original.call(connection, packet);
    }
}

File: src/main/kotlin/com/skysoft/mixin/MinecraftSlotLockMixin.kt
package com.skysoft.mixin

import com.llamalad7.mixinextras.injector.wrapoperation.Operation
import com.llamalad7.mixinextras.injector.wrapoperation.WrapOperation
import com.skysoft.features.inventory.SlotLockManager
import com.skysoft.utils.input.InputHandlingResult
import net.minecraft.client.Minecraft
import net.minecraft.client.multiplayer.ClientPacketListener
import net.minecraft.network.protocol.Packet
import net.minecraft.network.protocol.game.ServerboundPlayerActionPacket
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.injection.At

@Mixin(Minecraft::class)
open class MinecraftSlotLockMixin {
    @WrapOperation(
        method = ["handleKeybinds"],
        at = [
            At(
                value = "INVOKE",
                target = "Lnet/minecraft/client/multiplayer/ClientPacketListener;send(Lnet/minecraft/network/protocol/Packet;)V",
            ),
        ],
    )
    protected fun skysoftProtectLockedSlotsFromOffhandSwap(
        connection: ClientPacketListener,
        packet: Packet<*>,
        original: Operation<Void>,
    ) {
        val minecraft = this as Minecraft
        if (
            packet is ServerboundPlayerActionPacket &&
            packet.action == ServerboundPlayerActionPacket.Action.SWAP_ITEM_WITH_OFFHAND &&
            SlotLockManager.handleOffhandSwap(minecraft.player) == InputHandlingResult.CONSUMED
        ) {
            return
        }
        original.call(connection, packet)
    }
}
